src/dev/i18n/extract_html_messages.js

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { jsdom } from 'jsdom';
+import traverse from '@babel/traverse';
+import { parse } from '@babel/parser';
+import {
+  isConditionalExpression,
+  isDirectiveLiteral,
+  isObjectExpression,
+  isStringLiteral,
+} from '@babel/types';
+
+import {
+  isPropertyWithKey,
+  parseConditionalOperatorAST,
+  throwEntryException,
+} from './utils';
+import {
+  DEFAULT_MESSAGE_KEY,
+  ANGULAR_EXPRESSION_REGEX,
+  I18N_FILTER_IDS_REGEX,
+  I18N_FILTER_ARGS_REGEX,
+} from './constants';
+
+/**
+ * Extract default message from an angular filter expression argument
+ * @param {string} expression JavaScript code containing a filter object
+ * @returns {string} Default message
+ */
+function parseFilterObjectExpression(expression) {
+  let defaultMessage = '';
+
+  const filterObjectAST = parse(`+${expression}`); // parse an object expresssion instead of block statement
+  traverse(filterObjectAST, {
+    enter(path) {
+      if (isObjectExpression(path.node)) {
+        for (const property of path.node.properties) {
+          if (
+            isPropertyWithKey(property, DEFAULT_MESSAGE_KEY) &&
+            isStringLiteral(property.value)
+          ) {
+            defaultMessage = property.value.value;
+          }
+        }
+        path.stop();
+      }
+    },
+  });
+
+  return defaultMessage;
+}
+
+function parseConditionalOperatorExpression(expression) {
+  const ids = [];
+
+  traverse(parse(expression), {
+    enter(path) {
+      if (isDirectiveLiteral(path.node)) {
+        ids.push(path.node.value);
+        path.stop();
+      } else if (isConditionalExpression(path.node)) {
+        ids.push(...parseConditionalOperatorAST(path.node));
+        path.stop();
+      }
+    },
+  });
+
+  return ids;
+}
+
+function* getFilterMessages(htmlContent) {
+  const expressions = (htmlContent.match(ANGULAR_EXPRESSION_REGEX) || [])
+    .filter(
+      expression => expression.includes('|') && expression.includes('i18n')
```

Comment:
`"id" | i18n: { defaultMessage: "Default message" }`.
We can, but it will skip `|i18n` or `|<any whitespace bytes, e.g. "\n">i18n` syntax, which is allowed in angular expressions. Is it ok to disallow such syntax?

Reply:
Ideally I'd disallow this syntax and just check for `| i18n: {` and throw if someone tries to use `|i18n` or `|<any whitespace bytes, e.g. "\n">i18n`. But if that complicates things then let's keep it like it's with additional code comment regarding `|i18n` or `|<any whitespace bytes, e.g. "\n">i18n`.